include/sparrow/c_interface.hpp

```diff
@@ -0,0 +1,437 @@
+        }
+        schema->n_children = private_data->children_raw_ptr_vec_.size();
+        schema->children = private_data->children_raw_ptr_vec_.data();
+        schema->dictionary = private_data->dictionary_.get();
+        schema->release = delete_schema<Allocator>;
+        return schema;
+    };
+
+    struct ArrowArrayCustomDeleter
+    {
+        void operator()(ArrowArray* array) const
+        {
+            if (array->release != nullptr)
+            {
+                array->release(array);
+            }
+            delete array;
+        }
+    };
+
+    /**
+     * Creates an ArrowArray.
+     *
+     * @tparam T The type of the buffers of the ArrowArray
+     * @tparam Allocator The allocator for the buffers of the ArrowArray.
+     * @param length The logical length of the array (i.e. its number of items). Must be 0 or positive.
+     * @param null_count The number of null items in the array. May be -1 if not yet computed. Must be 0 or
+     * positive otherwise.
+     * @param offset The logical offset inside the array (i.e. the number of items from the physical start of
+     *               the buffers). Must be 0 or positive.
+     * @param n_buffers The number of physical buffers backing this array. The number of buffers is a
+     *                  function of the data type, as described in the Columnar format specification, except
+     *                  for the the binary or utf-8 view type, which has one additional buffer compared to the
+     *                  Columnar format specification (see Binary view arrays). Must be 0 or positive.
+     * @param children Vector of child arrays. Children must not be null.
+     * @param dictionary A pointer to the underlying array of dictionary values. Must be present if the
+     *                   ArrowArray represents a dictionary-encoded array. Must be null otherwise.
+     * @return The created ArrowArray.
+     */
+    template <class T, template <typename> class Allocator>
+        requires sparrow::allocator<Allocator<T>>
+    std::unique_ptr<ArrowArray, ArrowArrayCustomDeleter> make_array_constructor(
+        int64_t length,
+        int64_t null_count,
+        int64_t offset,
+        int64_t n_buffers,
+        std::vector<std::unique_ptr<ArrowArray>>&& children,
+        std::unique_ptr<ArrowArray> dictionary
+    )
+    {
+        SPARROW_ASSERT_TRUE(length >= 0)
+        SPARROW_ASSERT_TRUE(null_count >= -1)
+        SPARROW_ASSERT_TRUE(offset >= 0)
+        SPARROW_ASSERT_TRUE(n_buffers >= 0)
+        SPARROW_ASSERT_TRUE(std::ranges::none_of(
+            children,
+            [](const auto& child)
+            {
+                return child == nullptr;
+            }
+        ))
+
+        std::unique_ptr<ArrowArray, ArrowArrayCustomDeleter> array(new ArrowArray());
+
+        array->private_data = new ArrowArrayPrivateData<T, Allocator>(
+            std::move(children),
+            std::move(dictionary),
+            n_buffers,
+            length
```

Comment:
The buffers of an ArrowArray may not all have the same size, depending on the layout (which itself depends on the type). Therefore the enclosing function should probably take an additional parameter for the data type.